074-additional-volumes-support.md

```diff
@@ -0,0 +1,95 @@
+
+The specific usecase of ours, is that we want to set custom log4j logging for audit logs, to ensure that audit logs are *not* present in stdout, but can be picked up by a sidecar which uses OpenTelemtry or logstash to forward to the audit logs to a secure location.
+
+## Current situation
+
+The operator handles many configurations and operational aspects but lacks support for specifying additional volumes in the Custom Resource Definition (CRD). This limitation can restrict users from accessing log files or integrating with external storage solutions directly.
+
+## Motivation
+
+Adding support for additional volumes in the CRD will allow users to:
+
+ - Access log files for enhanced monitoring and debugging.
+ - Use additional configurations or secret management tools that require mounting extra volumes.
+
+Improve flexibility in managing Kafka deployments by allowing users to attach custom storage solutions based on their specific requirements.
+
+## Proposal
+
+### Considerations
+
+- CRD Changes: Update the Kafka CRDs to include an `additionalVolumes` field.
+- Validation: Ensure that the additional volumes are validated correctly and do not conflict with existing volume definitions such as `/var/lib/kafka/` and `/tmp`.
+- Documentation: Update the Strimzi documentation to include examples and guidelines on how to use the new `additionalVolumes` field.
+- Security: Ensure that the additional volumes do not introduce security vulnerabilities by validating the configurations.
+- Testing: Comprehensive testing to cover various scenarios and use cases, ensuring that the feature works as expected.
+
+### Possible Delivery Mechanisms
+
+- CRD Updates: Modify the existing CRDs to include the new field.
+  - This will be done by extending the `PodTemplate.java` class to include the new `additionalVolumes` field as a list.
+- Operator Logic: Update the operator logic to handle the additional volumes during the deployment process.
+  - This should be possible to implement in the `KafkaCluster.java` class.
+
+The configuration could look as follows:
+
+```yaml
+    additionalVolumes:
+    - name: example-configmap
+      path: /path/to/mount/volume
+      #subPath: mykey # Add this to mount only a specific key of the configmap/secret
+      configMap:
+        name: config-map-name
+    - name: temp
+      path: /tmp/logs
+      emptyDir: {}
+    - name: example-csi-volume
+      path: /path/to/mount/volume
+      #subPath: "subpath" # Add this to mount the CSI volume at a specific subpath
+      csi:
+        driver: csi-driver-name
+        readOnly: true
+        volumeAttributes:
+          secretProviderClass: example-secret-provider-class
+    - name: example-projected-volume
+      path: /path/to/mount/volume
+      projected:
+        sources:
+          serviceAccountToken
+```

```

Comment:
The template part of the API in general mirrors the Kubernetes API. So, why not do the same here as well:
* Add a `volumes` array in `PodTemplate` that mirrors the volume objects?
* The `Volume` API seems quite big and I have concerns about the CRD size especially during the transition to `v1` API in the future when multiple versions would need to be specified. Could you please add some considerations and analysis for that? Maybe it should be just defined as unstructured `Map<Stirng, Object>` to deal with the API size? 🤔
* Add `volumeMounts` to the `ContainerTemplate` to configure the mount points separately. I think this is important as there are other enhancement requests for things such as sidecars and those might need to share the volumes in some cases.

Reply:
> The template part of the API in general mirrors the Kubernetes API. So, why not do the same here as well:
> 
> * Add a `volumes` array in `PodTemplate` that mirrors the volume objects?
> * The `Volume` API seems quite big and I have concerns about the CRD size especially during the transition to `v1` API in the future when multiple versions would need to be specified. Could you please add some considerations and analysis for that? Maybe it should be just defined as unstructured `Map<Stirng, Object>` to deal with the API size? 🤔

A proposal would to not have an array of the direct `volumes`but something like this for the PodTemplate:
```
    private List<AdditionalVolume> additionalVolumes;
```

The new datatype would contain specific types we've selected:

```
    private SecretVolumeSource secret;
    private ConfigMapVolumeSource configMap;
    private EmptyDirVolumeSource emptyDir;
    private CSIVolumeSource csi;
```

This would also make the API size much smaller. I would like to avoid `Map<String, Object>` if possible

> * Add `volumeMounts` to the `ContainerTemplate` to configure the mount points separately. I think this is important as there are other enhancement requests for things such as sidecars and those might need to share the volumes in some cases.

Yeah that is fine. I have also talked to some other colleagues and I see operators such as Prometheus uses this approach.

Reply:
Not using directly the Kubernetes Volume API but having out own might be reasonable. I think that we will need way more than just the 4 you mentioned. But it might still save something and keep some control. And we might avoid supporting some of the deprecated in-tree driver volume types. But I think I personally would still call it `volumes`.

Reply:
> But I think I personally would still call it volumes.

I mean the field in the PodTemplate. The class name `AdditionalVolume` is of course fine.

Reply:
> `volumes` vs `additionalVolumes`
I think it is a bit misleading, and there will already be several other volumes present that is created automatically by the operator, but not visible in the CRD schema.

However, I have seen both naming convention used in other operators and I do not have a strong opinion here, so I am fine with getting inline with recommendations from the maintenance 😄